components/style/values/specified/image.rs

```diff
@@ -0,0 +1,370 @@
+    fn parse(input: &mut Parser) -> Result<Self, ()> {
+        match_ignore_ascii_case! { try!(input.expect_ident()),
+            "circle" => {
+                let position = input.try(Size::parse).unwrap_or(
+                    Size::Keyword(SizeKeyword::FarthestSide));
+                Ok(EndingShape::Circle(position))
+            },
+            "ellipse" => {
+              let (first_pos, second_pos) = if let Ok(first) = input.try(Size::parse) {
+                    if let Ok(second) = input.try(Size::parse) {
+                        (first, second)
+                    } else {
+                        (first, Size::Keyword(SizeKeyword::FarthestSide))
+                    }
+                } else {
+                    (Size::Keyword(SizeKeyword::FarthestSide),
+                     Size::Keyword(SizeKeyword::FarthestSide))
+                };
+                Ok(EndingShape::Ellipse(first_pos, second_pos))
+            },
+            _ => {
+                // If 1 <length> is present, it defaults to circle, otherwise defaults to ellipse.
+                if let Ok(first) = input.try(Size::parse) {
+                    if let Ok(second) = input.try(Size::parse) {
+                        Ok(EndingShape::Ellipse(first, second))
+                    } else {
+                        if let Size::Length(length) = first {
+                            if let LengthOrPercentage::Length(len) = length {
+                                Ok(EndingShape::Circle(Size::Length(len)))
+                            } else {
+                                Err(())
+                            }
+                        } else {
+                            Err(())
+                        }
+                    }
+                } else {
+                    Err(())
+                }
+            }
+        }
+    }
+}
+
+impl ToCss for EndingShape {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        match *self {
+            EndingShape::Circle(ref length) => {
+                try!(dest.write_str("circle "));
+                try!(length.to_css(dest));
+            },
+            EndingShape::Ellipse(ref first_len, ref second_len) => {
+                try!(dest.write_str("ellipse "));
+                try!(first_len.to_css(dest));
+                try!(dest.write_str(" "));
+                try!(second_len.to_css(dest));
+            },
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Size<T: Clone + Parse + ToCss + ToComputedValue> {
+    Length(T),
+    Keyword(SizeKeyword),
+}
+
+impl<T: Clone + Parse + ToCss + ToComputedValue> Size<T> {
+    pub fn parse(input: &mut Parser) -> Result<Size<T>, ()> {
+        if let Ok(keyword) = input.try(SizeKeyword::parse) {
+            Ok(Size::Keyword(keyword))
+        } else {
+            let length = input.try(<T>::parse).unwrap(); // TODO(canaltinova): It may not work
```

Comment:
This is better written as `try!(Size::Length(input.try(Parse::parse)))` in one line 😄